lib/shared/src/misc/asyncGenerator.test.ts

```diff
@@ -0,0 +1,181 @@
+            }
+        }
+
+        const abortController = new AbortController()
+        const result = await firstValueFrom(infiniteGenerator(abortController.signal), abortController)
+        expect(result).toBe(0)
+        expect(abortController.signal.aborted).toBe(true)
+    })
+})
+
+describe('combineLatest', () => {
+    test('combine latest values', async () => {
+        const g = combineLatest([asyncGeneratorWithValues(1, 2), asyncGeneratorWithValues('a', 'b')])
+        const { values, done } = readValuesFrom(g)
+        await done
+        expect(values).toEqual<typeof values>([
+            [1, 'a'],
+            [2, 'a'],
+            [2, 'b'],
+        ])
+    })
+
+    test('handles empty input', async () => {
+        const combined = combineLatest([] as any)
+        const results = []
+
+        for await (const result of combined) {
+            results.push(result)
+        }
+
+        expect(results).toEqual([])
+    })
+
+    test('keeps going after one returns', { timeout: 1000 }, async () => {
+        const returnsAfter1 = async function* () {
+            yield 1
+        }
+        const returnsAfterC = async function* () {
+            yield 'a'
+            yield 'b'
+            await new Promise(resolve => setTimeout(resolve))
+            yield 'c'
+        }
+
+        const g = combineLatest([returnsAfter1(), returnsAfterC()])
+        const { values, done } = readValuesFrom(g)
+        await done
+        expect(values).toEqual<typeof values>([
+            [1, 'a'],
+            [1, 'b'],
+            [1, 'c'],
+        ])
+    })
+
+    test('aborts', async () => {
+        const abortController = new AbortController()
+        const g = combineLatest([asyncGeneratorWithValues(1, 2)], abortController.signal)
+        const { values, done } = readValuesFrom(g)
+        abortController.abort()
+        await done
+        expect(values).toEqual<typeof values>([])
+    })
+})
+
+describe('memoizeLatest', () => {
+    test('memoize latest values', async () => {
+        let calls = 0
+        const factory = async function* (n: number, signal?: AbortSignal) {
+            calls++
+            await new Promise(resolve => setTimeout(resolve))
+            yield n
+            yield n + 1
+        }
+
+        const memoized = memoizeLastValue(factory, args => args[0])
+        expect(calls).toBe(0)
+
+        const signal = new AbortController().signal
+        expect(await allValuesFrom(memoized(0, signal))).toEqual<number[]>([0, 1])
+        expect(calls).toBe(1)
+
+        expect(await allValuesFrom(memoized(10, signal))).toEqual<number[]>([10, 11])
+        expect(calls).toBe(2)
+
+        expect(await allValuesFrom(memoized(0, signal))).toEqual<number[]>([1, 0, 1])
+        expect(calls).toBe(3)
+    })
+})
+
+describe('omitDuplicateSequences', () => {
+    test('should omit duplicate consecutive values', async () => {
+        const input = asyncGeneratorWithValues(1, 1, 2, 2, 2, 3, 4, 4, 5)
```

Comment:
Worth testing that non-consecutive duplicate values are not filtered

```suggestion
        const input = asyncGeneratorWithValues(1, 1, 2, 2, 2, 1, 3, 4, 4, 5)
```